src/nix/run.cc

```diff
@@ -20,45 +21,43 @@ using namespace nix;
 
 std::string chrootHelperName = "__run_in_chroot";
 
-struct RunCommon : virtual Command
-{
+namespace nix {
 
-    using Command::run;
+void runProgramInStore(ref<Store> store,
+    const std::string & program,
+    const Strings & args)
+{
+    stopProgressBar();
 
-    void runProgram(ref<Store> store,
-        const std::string & program,
-        const Strings & args)
-    {
-        stopProgressBar();
+    restoreProcessContext();
 
-        restoreProcessContext();
+    /* If this is a diverted store (i.e. its "logical" location
+       (typically /nix/store) differs from its "physical" location
+       (e.g. /home/eelco/nix/store), then run the command in a
+       chroot. For non-root users, this requires running it in new
+       mount and user namespaces. Unfortunately,
+       unshare(CLONE_NEWUSER) doesn't work in a multithreaded program
+       (which "nix" is), so we exec() a single-threaded helper program
+       (chrootHelper() below) to do the work. */
```

Comment:
This sounds like it would also fix #4337 for `nix develop`. Is it the case?

Reply:
I think so.